<a href="https://colab.research.google.com/github/TheGodseeker/ML_PR/blob/main/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv("/content/data/anime.csv")
df

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 7 columns]

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [ ]:
#Колонки
df.columns

Index(['anime_id', 'name', 'genre', 'type', 'episodes', 'rating', 'members'], dtype='object')

In [ ]:
#Первые 5 значений
df[:5]

anime_id                              name  \
0     32281                    Kimi no Na wa.   
1      5114  Fullmetal Alchemist: Brotherhood   
2     28977                          Gintama°   
3      9253                       Steins;Gate   
4      9969                     Gintama&#039;   

                                               genre   type episodes  rating  \
0               Drama, Romance, School, Supernatural  Movie        1    9.37   
1  Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64    9.26   
2  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.25   
3                                   Sci-Fi, Thriller     TV       24    9.17   
4  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.16   

   members  
0   200630  
1   793665  
2   114262  
3   673572  
4   151266

In [ ]:
#Последние 5 значений
df[-5:]

anime_id                                               name   genre  \
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami  Hentai   
12290      5543                                        Under World  Hentai   
12291      5621                     Violence Gekiga David no Hoshi  Hentai   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...  Hentai   
12293     26081                   Yasuji no Pornorama: Yacchimae!!  Hentai   

        type episodes  rating  members  
12289    OVA        1    4.15      211  
12290    OVA        1    4.28      183  
12291    OVA        4    4.88      219  
12292    OVA        1    4.98      175  
12293  Movie        1    5.46      142

In [ ]:
#Статистику
df.describe(include="all")

anime_id                     name   genre   type episodes  \
count   12294.000000                    12294   12232  12269    12294   
unique           NaN                    12292    3264      6      187   
top              NaN  Shi Wan Ge Leng Xiaohua  Hentai     TV        1   
freq             NaN                        2     823   3787     5677   
mean    14058.221653                      NaN     NaN    NaN      NaN   
std     11455.294701                      NaN     NaN    NaN      NaN   
min         1.000000                      NaN     NaN    NaN      NaN   
25%      3484.250000                      NaN     NaN    NaN      NaN   
50%     10260.500000                      NaN     NaN    NaN      NaN   
75%     24794.500000                      NaN     NaN    NaN      NaN   
max     34527.000000                      NaN     NaN    NaN      NaN   

              rating       members  
count   12064.000000  1.229400e+04  
unique           NaN           NaN  
top              NaN           NaN  
freq             NaN           NaN  
mean        6.473902  1.807134e+04  
std         1.026746  5.482068e+04  
min         1.670000  5.000000e+00  
25%         5.880000  2.250000e+02  
50%         6.570000  1.550000e+03  
75%         7.180000  9.437000e+03  
max        10.000000  1.013917e+06

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





**Ответ 1:** rating=6.473902; members=1.807134e+04

**Ответ 2:** минимальное = 1; максимальное = 34527



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [ ]:
df.genre = df.genre.astype("category")
df.rating = df.rating.astype("Float64")
df.anime_id = df.anime_id.astype("Int64")

#Вывести типы колонок тут
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   anime_id  12294 non-null  Int64   
 1   name      12294 non-null  object  
 2   genre     12232 non-null  category
 3   type      12269 non-null  object  
 4   episodes  12294 non-null  object  
 5   rating    12064 non-null  Float64 
 6   members   12294 non-null  int64   
dtypes: Float64(1), Int64(1), category(1), int64(1), object(3)
memory usage: 779.0+ KB


Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [ ]:
#Фильтр колонки 1
df[df["genre"]== "Hentai"]

anime_id                                               name   genre  \
11114     11879                                Oni Chichi: Re-born  Hentai   
11115     29575                                 Mankitsu Happening  Hentai   
11116     15843                       Koiito Kinenbi The Animation  Hentai   
11117     21097                                Oni Chichi: Rebuild  Hentai   
11119     10779                Eroge! H mo Game mo Kaihatsu Zanmai  Hentai   
...         ...                                                ...     ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami  Hentai   
12290      5543                                        Under World  Hentai   
12291      5621                     Violence Gekiga David no Hoshi  Hentai   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...  Hentai   
12293     26081                   Yasuji no Pornorama: Yacchimae!!  Hentai   

        type episodes  rating  members  
11114    OVA        2    7.89    14342  
11115    OVA        4    7.83     8510  
11116    OVA        2    7.75     6940  
11117    OVA        3    7.75     9825  
11119    OVA        6    7.68    20316  
...      ...      ...     ...      ...  
12289    OVA        1    4.15      211  
12290    OVA        1    4.28      183  
12291    OVA        4    4.88      219  
12292    OVA        1    4.98      175  
12293  Movie        1    5.46      142  

[823 rows x 7 columns]

In [ ]:
#Фильтр колонки 2
df[df["rating"] >= 8.5]

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
10616     28023                                  Tsuru no Sugomori   
10718     22219                               Warera Salaryman Tou   
10786     26313                    Yakusoku: Africa Mizu to Midori   
10793     28557  Yamete! Writer Asobi: Doubutsu Mura no Shoubou...   
10847     26097                                    Yume no Tsuzuki   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
10616                                               Kids  Movie        1   
10718                              Comedy, Slice of Life     TV       26   
10786                                        Drama, Kids    OVA        1   
10793                                        Drama, Kids    OVA        1   
10847                                        Drama, Kids    OVA        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
10616     8.5       46  
10718    8.67       57  
10786    9.25       53  
10793    8.67       40  
10847    8.67       53  

[137 rows x 7 columns]

In [ ]:
#Фильтр колонки 3
df[df["members"] < 10000]

anime_id                                               name  \
111        2921                                    Ashita no Joe 2   
205       32005    Detective Conan Movie 20: The Darkest Nightmare   
261        1382                          Yuusha-Ou GaoGaiGar Final   
302       29893                                  Aria The Avvenire   
305        3371  Ginga Eiyuu Densetsu Gaiden: Senoku no Hoshi, ...   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                          genre     type episodes  rating  \
111                               Drama, Sports       TV       47    8.48   
205            Action, Mystery, Police, Shounen    Movie        1    8.32   
261    Adventure, Drama, Mecha, Sci-Fi, Shounen      OVA        8    8.25   
302     Fantasy, Sci-Fi, Shounen, Slice of Life  Special        3     8.2   
305             Action, Military, Sci-Fi, Space      OVA       24     8.2   
...                                         ...      ...      ...     ...   
12289                                    Hentai      OVA        1    4.15   
12290                                    Hentai      OVA        1    4.28   
12291                                    Hentai      OVA        4    4.88   
12292                                    Hentai      OVA        1    4.98   
12293                                    Hentai    Movie        1    5.46   

       members  
111       6259  
205       6229  
261       5001  
302       9500  
305       8621  
...        ...  
12289      211  
12290      183  
12291      219  
12292      175  
12293      142  

[9304 rows x 7 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [ ]:
#Группировка
group = df.groupby(["type"])
group.count()

anime_id  name  genre  episodes  rating  members
type                                                     
Movie        2348  2348   2306      2348    2297     2348
Music         488   488    488       488     488      488
ONA           659   659    655       659     652      659
OVA          3311  3311   3310      3311    3285     3311
Special      1676  1676   1674      1676    1671     1676
TV           3787  3787   3777      3787    3671     3787

In [ ]:
#Среднее значение
group.mean()


<ipython-input-56-fcb56ce3761e>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  group.mean()
<ipython-input-56-fcb56ce3761e>:3: FutureWarning: The default value of numeric_only in DataFrameGroupBy.median is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  group.median()


anime_id  rating  members
type                              
Movie     10981.5    6.47    499.5
Music     24902.0   5.625    227.5
ONA       27911.0   5.755    312.0
OVA        6690.0    6.38   1663.0
Special   15324.0    6.63   2246.5
TV         7525.0    6.94   5947.0

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация

count = 0
for i in df[df["type"] == "Movie"]["name"]: count += 1
print("Count names of movies =",count)

Count names of movies = 2348


In [ ]:
# Cр. значения

sum = 0
count = 0
for i in df[df["type"] == "Movie"]["anime_id"]:
  sum += i
  count += 1
print("Ср. значение anime_id у Movie =", (sum/count))

sum = 0
count = 0
for i in df[df["type"] == "Music"]["anime_id"]:
  sum += i
  count += 1
print("Ср. значение anime_id у Music =", (sum/count))

sum = 0
count = 0
for i in df[df["type"] == "ONA"]["anime_id"]:
  sum += i
  count += 1
print("Ср. значение anime_id у ONA =", (sum/count))

sum = 0
count = 0
for i in df[df["type"] == "OVA"]["anime_id"]:
  sum += i
  count += 1
print("Ср. значение anime_id у OVA =", (sum/count))

sum = 0
count = 0
for i in df[df["type"] == "Special"]["anime_id"]:
  sum += i
  count += 1
print("Ср. значение anime_id у Special =", (sum/count))

sum = 0
count = 0
for i in df[df["type"] == "Music"]["anime_id"]:
  sum += i
  count += 1
print("Ср. значение anime_id у TV =", (sum/count))

Ср. значение anime_id у Movie = 14898.067291311754
Ср. значение anime_id у Music = 22445.54713114754
Ср. значение anime_id у ONA = 23283.47192716237
Ср. значение anime_id у OVA = 11322.082150407732
Ср. значение anime_id у Special = 16666.27983293556
Ср. значение anime_id у TV = 22445.54713114754


In [ ]:
#Медиана
#print(df[df["type"] == "Movie"]["anime_id"])

#print("\n\n",int(df[df["type"] == "Movie"]["anime_id"][1:2]))
#print("\n\n",int(df[df["type"] == "Movie"]["anime_id"][2:3]))

count = 0
for i in df[df["type"] == "Movie"]["anime_id"]: count += 1
half = int(count / 2)
if(count % 2 != 0):
  print("Медиана anime_id у Movie =", int(df[df["type"] == "Movie"]["anime_id"][half:half+1]))
else:
  med = int(df[df["type"] == "Movie"]["anime_id"][half:half+1])+int(df[df["type"] == "Movie"]["anime_id"][half+1:half+2])
  med /= 2
  print("Медиана anime_id у Movie =", med)

count = 0
for i in df[df["type"] == "Music"]["anime_id"]: count += 1
half = int(count / 2)
if(count % 2 != 0):
  print("Медиана anime_id у Music =", int(df[df["type"] == "Music"]["anime_id"][half:half+1]))
else:
  med = int(df[df["type"] == "Music"]["anime_id"][half:half+1])+int(df[df["type"] == "Music"]["anime_id"][half+1:half+2])
  med /= 2
  print("Медиана anime_id у Music =", med)

count = 0
for i in df[df["type"] == "ONA"]["anime_id"]: count += 1
half = int(count / 2)
if(count % 2 != 0):
  print("Медиана anime_id у ONA =", int(df[df["type"] == "ONA"]["anime_id"][half:half+1]))
else:
  med = int(df[df["type"] == "ONA"]["anime_id"][half:half+1])+int(df[df["type"] == "ONA"]["anime_id"][half+1:half+2])
  med /= 2
  print("Медиана anime_id у ONA =", med)

count = 0
for i in df[df["type"] == "OVA"]["anime_id"]: count += 1
half = int(count / 2)
if(count % 2 != 0):
  print("Медиана anime_id у OVA =", int(df[df["type"] == "OVA"]["anime_id"][half:half+1]))
else:
  med = int(df[df["type"] == "OVA"]["anime_id"][half:half+1])+int(df[df["type"] == "OVA"]["anime_id"][half+1:half+2])
  med /= 2
  print("Медиана anime_id у OVA =", med)

count = 0
for i in df[df["type"] == "Special"]["anime_id"]: count += 1
half = int(count / 2)
if(count % 2 != 0):
  print("Медиана anime_id у Special =", int(df[df["type"] == "Special"]["anime_id"][half:half+1]))
else:
  med = int(df[df["type"] == "Special"]["anime_id"][half:half+1])+int(df[df["type"] == "Special"]["anime_id"][half+1:half+2])
  med /= 2
  print("Медиана anime_id у Special =", med)

count = 0
for i in df[df["type"] == "TV"]["anime_id"]: count += 1
half = int(count / 2)
if(count % 2 != 0):
  print("Медиана anime_id у TV =", int(df[df["type"] == "TV"]["anime_id"][half:half+1]))
else:
  med = int(df[df["type"] == "TV"]["anime_id"][half:half+1])+int(df[df["type"] == "TV"]["anime_id"][half+1:half+2])
  med /= 2
  print("Медиана anime_id у TV =", med)

Медиана anime_id у Movie = 30677.0
Медиана anime_id у Music = 28496.5
Медиана anime_id у ONA = 33871
Медиана anime_id у OVA = 29273
Медиана anime_id у Special = 10280.0
Медиана anime_id у TV = 1893
